### PR TITLE
fixed 5602, expose cached build script output in json output

### LINKED ISSUE #5602: Expose cached build script output in the json output
Let's say I have a hello world app with the following `build.rs`:

```rust
fn main() {
    println!("cargo:rustc-cfg=foo");
}
```

Running `cargo build --quiet --message-format=json` on a cold cache yields the following output, containing the `cfg=foo` from the build script:

```
{"features":[],"filenames":["/home/konsti/build-script-json/target/debug/build/build-script-json-a721905d0b722f88/build-script-build"],"fresh":false,"package_id":"build-script-json 0.1.0 (path+file:///home/konsti/build-script-json)","profile":{"debug_assertions":true,"debuginfo":2,"opt_level":"0","overflow_checks":true,"test":false},"reason":"compiler-artifact","target":{"crate_types":["bin"],"kind":["custom-build"],"name":"build-script-build","src_path":"/home/konsti/build-script-json/build.rs"}}
{"cfgs":["foo"],"env":[],"linked_libs":[],"linked_paths":[],"package_id":"build-script-json 0.1.0 (path+file:///home/konsti/build-script-json)","reason":"build-script-executed"}
{"features":[],"filenames":["/home/konsti/build-script-json/target/debug/build-script-json"],"fresh":false,"package_id":"build-script-json 0.1.0 (path+file:///home/konsti/build-script-json)","profile":{"debug_assertions":true,"debuginfo":2,"opt_level":"0","overflow_checks":true,"test":false},"reason":"compiler-artifact","target":{"crate_types":["bin"],"kind":["bin"],"name":"build-script-json","src_path":"/home/konsti/build-script-json/src/main.rs"}}
```

Running the same command another time will omit the build script output:

```
{"features":[],"filenames":["/home/konsti/build-script-json/target/debug/build/build-script-json-a721905d0b722f88/build-script-build"],"fresh":true,"package_id":"build-script-json 0.1.0 (path+file:///home/konsti/build-script-json)","profile":{"debug_assertions":true,"debuginfo":2,"opt_level":"0","overflow_checks":true,"test":false},"reason":"compiler-artifact","target":{"crate_types":["bin"],"kind":["custom-build"],"name":"build-script-build","src_path":"/home/konsti/build-script-json/build.rs"}}
{"features":[],"filenames":["/home/konsti/build-script-json/target/debug/build-script-json"],"fresh":true,"package_id":"build-script-json 0.1.0 (path+file:///home/konsti/build-script-json)","profile":{"debug_assertions":true,"debuginfo":2,"opt_level":"0","overflow_checks":true,"test":false},"reason":"compiler-artifact","target":{"crate_types":["bin"],"kind":["bin"],"name":"build-script-json","src_path":"/home/konsti/build-script-json/src/main.rs"}}
```

While a file with the cached output exists, the path to this file not part of json output. That makes it impossible to determine which configuration is set from the json output without rebuilding each time.

For the real-world context: I want to package libraries build with [pyo3](https://github.com/PyO3/pyo3) and need the build script output to determine the python variant used.

### PR DESCRIPTION
**Pull request by debris**
fixes #5602

### CODE REVIEW

File: tests/testsuite/build.rs [hunk 1]
```diff
@@ -3976,11 +3976,13 @@ fn compiler_json_error_format() {
             name = "foo"
             version = "0.5.0"
             authors = ["wycats@example.com"]
+            build = "build.rs"
```
Comment (matklad) [inline tests/testsuite/build.rs]: This line is not neccessary, `build.rs` is the default name for quite some time now.

File: src/cargo/core/compiler/custom_build.rs [hunk 2]
```diff
@@ -365,6 +365,22 @@ fn build_work<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>) -> CargoRes
...
+                .map(|l| l.display().to_string())
+                .collect::<Vec<_>>();
+            machine_message::emit(&machine_message::BuildScript {
+                package_id: &id,
+                linked_libs: &output.library_links,
+                linked_paths: &library_paths,
+                cfgs: &output.cfgs,
+                env: &output.env,
+            });
+        }
```
Comment (matklad) [inline src/cargo/core/compiler/custom_build.rs]: Let's extract this chunk of code to a function, to avoid duplication?

**Comment by matklad** [comment]
@bors r+

Thanks!

**Comment by debris** [comment]
thanks for the review @matklad !

**Comment by konstin** [comment]
Thank you!

**Pull request merged**